language/move-prover/bytecode-to-boogie/src/spec_translator.rs

```diff
@@ -602,29 +608,40 @@ impl<'a> SpecTranslator<'a> {
         } else {
             false
         };
-        if let SignatureToken::Struct(handle_idx, type_args) = sig {
-            let struct_handle = self.module.struct_handle_at(*handle_idx);
-            // We cannot lookup type information for fields of imported modules via the handle view.
-            // Instead we need to retrieve the struct definition, which might be in another module.
-            // TODO: validate that I've not missed some feature in file_format.rs. It might be
-            //   considered an issue in the byte code representation that we can't do this; it
-            //   might be better that bytecode fully specifies all metadata about dependencies.
-            if let Some(struct_def) = self.get_struct_definition(struct_handle) {
+
+        if let Some(struct_handle_view) = SignatureTokenView::new(self.module, sig).struct_handle()
+        {
+            // NOTE: this returns a StructDefinitionViewFrom a different module!
+            if let Some(struct_def) = self.get_struct_definition(struct_handle_view.handle()) {
                 let struct_name = struct_def.name().to_string(); // because of borrowing
                 if let Some(field_def) = self.get_field_definition(&struct_def, field) {
-                    let field_type = field_def.signature_token().substitute(type_args);
```

Comment:
This was the bug. This is a `SignatureToken` from the module where the field is declared, but we need one for the current module.

Reply:
I think a simpler solution which avoids those 'ghost' types is to just define our own SignatureToken type which is using a StructHandle but is otherwise the same. It would not require those 'ghost' types. But I'm good with submitting this fix, and can take care of this different solution.

Reply:
Good point. I think if the verifier used `TypeTag`s instead of `SignatureToken`s to record the types of (e.g.) Boogie expressions, that would be the simplest/least error-prone approach. Worth discussing/exploring in the future.